File: OptimProblemDMMOP.m
% This class creates the object problem. CEC'2013 test problems have already been defined with ID=1,...,20. It is possible to add custom problems 
% which is shown by one example.
% Written by Ali Ahrari (aliahrari1983@gmail.com)
% last updated by Ali Ahrari on 11 Jan 2021""" 



classdef OptimProblemDMMOP < OptimProblem
    % The main class. It creates an object which determine the problem 
    properties 
        extProb % everything related to the problem subclass (external problem)
    end
    methods
        function problem=OptimProblemDMMOP(pid) % constructor
            problem@OptimProblem(pid);
            problem.suite='DMMOP';
            problem.isDynamic=true;
            problem.extProb = DMMOP(pid);
            problem.PID = pid; % function ID
        end

        function set_problem_data(problem) % Calculate and set the problem information --> must be defined 
            % set the search bounds  
            problem.lowBound=problem.extProb.lower;
            problem.upBound=problem.extProb.upper;
            problem.dim=problem.extProb.D;
            problem.maxEval=problem.extProb.evaluation; 
            problem.globMinData.Ngmin=size(problem.extProb.o,1);
        end
        function f=func_eval(problem,x) 
            % Objective function: Evaluate the solution x 
            % x: a 1-D array representing the solution to be evaluated 
            f = problem.extProb.GetFits(x);
            if numel(f)==0
                f=inf;
            else
                f=-f;
                problem.numCallF=problem.numCallF+1;
            end
            problem.numCallF=problem.extProb.evaluated;
        end
    end
end

File: optim_problem_dmmop.py
"""Optimization problem for DMMOP (CEC 2022 dynamic benchmark)."""
from __future__ import annotations
import numpy as np
import sys
from pathlib import Path

# Add dmmops_py to path
sys.path.insert(0, str(Path(__file__).parent / 'dmmops_py'))

from optim_problem import OptimProblem
from data_structures import GlobalMinimaData
from dmmops_py.problem.dmmop import DMMOP


class OptimProblemDMMOP(OptimProblem):
    """Optimization problem using DMMOP external benchmark."""
    
    def __init__(self, function_number: int):
        """
        Initialize DMMOP problem.
        
        Args:
            function_number: DMMOP function number (1-24)
        
        Note: Dimension is determined automatically by get_info() based on function number.
        """
        self.suite = 'DMMOP'
        self.isDynamic = True
        self.PID = function_number
        
        # Initialize DMMOP problem (dimension is set by get_info)
        self.extProb = DMMOP(function_number)
        
        # Set problem parameters
        self.lowBound = self.extProb.lower
        self.upBound = self.extProb.upper
        self.dim = self.extProb.D
        self.maxEval = self.extProb.evaluation
        
        # Initialize from parent
        self.numCallF = 0
        self.knownChanges = True
        self.globMinData = GlobalMinimaData()
        self.globMinData.Ngmin = self.extProb.o.shape[0] if self.extProb.o is not None else 0

    def func_eval(self, x: np.ndarray) -> float | np.ndarray:
        """Objective function: Evaluate using DMMOP."""
        f = self.extProb.GetFits(x if x.ndim == 2 else x[None, :])
        if f.size == 0:
            return np.inf if x.ndim == 1 else np.full(x.shape[0], np.inf)
        f = -f  # DMMOP returns negative fitness
        self.numCallF = self.extProb.evaluated
        return f[0] if x.ndim == 1 and f.size == 1 else f
